src/main/java/edu/ucar/unidata/rosetta/service/DataManagerImpl.java

```diff
@@ -0,0 +1,283 @@
+            /*
+            NOTE: code in the try block actually throws a bunch of different exceptions, including
+            java.lang.Exception itself.  Hence, the use catch of the generic Exception class to
+            catch them all (otherwise I normally would not catch with just java.lang.Exception).
+             */
+            throw new RosettaDataException("Unable to populate data object by reflection: " + e);
+        }
+        return persistedData;
+    }
+
+    /**
+     * Converts .xls and .xlsx files to .csv files.
+     *
+     * @param id        The unique id associated with the file (a subdir in the uploads directory).
+     * @param fileName  The name of the .xls or .xlsx file to convert.
+     * @return          The name of the converted .csv file.
+     * @throws IOException  If unable to convert to .csv file.
+     */
+    @Override
+    public String convertToCSV(String id, String fileName) throws IOException {
+        String extension = FilenameUtils.getExtension(fileName);
+        if (!(extension.equals("xls") || extension.equals("xlsx")))
+            throw new UnsupportedDataTypeException("Attempting to convert a non .xls type file to .csv format.");
+
+        String xlsFilePath = FilenameUtils.concat(FilenameUtils.concat(getUploadDir(), id), fileName);
+        // Change the file on disk.
+        boolean conversionSuccessful = XlsToCsv.convert(xlsFilePath, null);
+        String csvFileName = null;
+        if (conversionSuccessful)
+            csvFileName = FilenameUtils.removeExtension(fileName) + ".csv";
+        else
+            throw new IOException("Attempting to convert a non .xls type file to .csv format.");
+
+        return csvFileName;
+    }
+
+    /**
+     * Creates a subdirectory in the designated uploads directory using the (unique) id
+     * and writes the given file to the uploads subdirectory.
+     *
+     * @param id        The unique id associated with the file (a subdir in the uploads directory).
+     * @param fileName  The name of the file to save to disk.
+     * @param file      The CommonsMultipartFile to save to disk.
+     * @throws SecurityException  If unable to write file to disk because of a JVM security manager violation.
+     * @throws IOException  If unable to write file to disk.
+     */
+    @Override
+    public void writeUploadedFileToDisk(String id, String fileName, CommonsMultipartFile file) throws SecurityException, IOException {
+        String filePath = FilenameUtils.concat(getUploadDir(), id);
+        File localFileDir = new File(filePath);
+
+        if (!localFileDir.exists())
+            if (!localFileDir.mkdirs())
+                throw new IOException("Unable to create " + id + " subdirectory in uploads directory.");
+
+
+        logger.info("Writing uploaded file " + fileName + " to disk");
+        File uploadedFile = new File(FilenameUtils.concat(filePath, fileName));
+
+        FileOutputStream outputStream = new FileOutputStream(uploadedFile);
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "FileOutputStream" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2095)